### PR TITLE
Flip flop attachment pod

### PR DESCRIPTION
**Pull request by awels**
**What this PR does / why we need it**:
The PR modifies the hotplug code to recreate the attachment pod with the current number of hotplug volumes in it. Instead of one pod per volume. This reduces the number of extra pods used by the hotplug system to 1 instead of N.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Enhancement: remove one attachment pod per disk limit
```

### CODE REVIEW

**Comment by awels** [comment]
@davidvossel Hey, can you take a look at this PR, it implements the single pod for all hotplugged disks. @vasiliy-ul is using this PR for his Live Migration PR.

**Review by davidvossel — COMMENTED** [review]
@awels looking through this, one thing that's not immediately obvious to me is how the transition from the old attachment pod to the new attachment pod takes place seamlessly. I want to make sure my understanding is accurate. 

For block devices, it seems like we were finding the underlying device on the host and passing that into the launcher pod. I think i get how that works with the flip flop pod since i'm guessing a consistent device stays put on the host as long as one pod on the host has the pvc attached.

However, for file based PVCs, is the underlying mount point being used consistent as new pods with the PVC are attached and old ones are deleted?

Also, for ReadWriteOnce or ReadOnly, is the guarantee that we can attach those pvcs to multiple pods as long as they're on the same host?  If so, is this part of the CSI spec? I just want to make sure this behavior is something we can depend on.

**Comment by awels** [comment]
There is a very big difference between mounting and attaching. So the way the new flip flop pods is that any new volumes are both attached to a node and mounted into the attachment pod. This is the same as the current many attachment pod approach. However when we flip flop because a new disk is added, any volumes that are ready (and thus in use by the VM) are only attached, not mounted in the new attachment pod. You can see this in the attachment pod spec where only the volume is defined, but it is not defined in the container section (the part that gives it a path inside a container). Doing it this way causes the CSI system to attach the volume to the node (if not attached already) but not to attempt to mount it inside the pod. 

We were having issues with for instance rook-ceph-rbd RWO filesystem volumes, where attempting to mount on the second pod would fail because there was a RW file system mounted somewhere else (in the VM). At first we attempted a read only mount in the flip flop pod, but some provisioners don't support read only volumes (in particular longhorn). But we then realized the real purpose of the attachment pods is to attach the volume to the node and have a reference that can properly clean up the attachment, but we don't really care if the volume is mounted in that pod once the VM is using it.

As for your question, we are not guaranteeing anything about mounting the volume into multiple pods, we are only guaranteeing that the volume is attached to the same node the VM is running on.

**Comment by davidvossel** [comment]
> , any volumes that are ready (and thus in use by the VM) are only attached, not mounted in the new attachment pod.

@awels i see. are you sure the old pod can successfully terminate in this case then? It's possible the kubelet can't clean up the pod until the PVC mount point is no longer in use for that pod. 

Said another way, after launching a new flip/flop pod, does the old pod actually exit the "terminating" state or is it stuck in "terminating" indefinitely until the VMI shuts down?

File: tests/storage/hotplug.go [hunk 1]
```diff
@@ -829,48 +829,6 @@ var _ = SIGDescribe("Hotplug", func() {
 				vmi = tests.RunVMIAndExpectLaunch(vmi, 240)
 			})
 
-			It("should mark VMI failed, if an attachment pod is deleted", func() {
```
Comment (davidvossel) [inline tests/storage/hotplug.go:832]: we need to add a functional test that validates that the old attachment pods exit as expected as new pods are created during hot plug

Comment (awels) [reply tests/storage/hotplug.go:832]: Sure I can add such a test.

**Comment by awels** [comment]
> > , any volumes that are ready (and thus in use by the VM) are only attached, not mounted in the new attachment pod.
> 
> @awels i see. are you sure the old pod can successfully terminate in this case then? It's possible the kubelet can't clean up the pod until the PVC mount point is no longer in use for that pod.
> 
> Said another way, after launching a new flip/flop pod, does the old pod actually exit the "terminating" state or is it stuck in "terminating" indefinitely until the VMI shuts down?

Yes I am positive this happens, in fact usually the old attachment pod is gone before the new one is started, so there is a period with no attachment pods (the attached volumes cannot be detached due to being in use by the VM) the attachment pod is purely so the kubelet can properly detach volumes from the node that get removed from the VM

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

impressive stuff :)  I had one minor comment we might want to follow up on in a separate PR.

Moving forward with what we have now, we just need to be aware that this change breaks update compatibility for people using HotPlug with a previous version of KubeVirt due to the mount path changes.

We'll need to make sure to alert the community that a backwards incompatible change is occuring. Since HotPlug is feature gated and subject to change anyway, i don't think this is a big deal. THe benefit of this PR outweighs maintaining backwards compatibility here.

**Pull request merged**